File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sinks/StreamSelectTableSink.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.planner.sinks

import org.apache.flink.api.common.ExecutionConfig
import org.apache.flink.core.execution.JobClient
import org.apache.flink.streaming.api.datastream.{DataStream, DataStreamSink}
import org.apache.flink.streaming.experimental.{CollectSink, SocketStreamIterator}
import org.apache.flink.table.api.TableSchema
import org.apache.flink.table.api.internal.SelectTableSink
import org.apache.flink.table.sinks.AppendStreamTableSink
import org.apache.flink.table.types.DataType
import org.apache.flink.types.Row

import java.net.InetAddress
import java.util

/**
  * A [[SelectTableSink]] for streaming select job.
  *
  * <p><strong>NOTES:</strong> This is a temporary solution,
  * once FLINK-14807 is finished, the implementation should be changed.
  * Currently, only insert changes (AppendStreamTableSink) is supported.
  * Once FLINK-16998 is finished, all kinds of changes will be supported.
  */
class StreamSelectTableSink(tableSchema: TableSchema)
  extends AppendStreamTableSink[Row]
  with SelectTableSink {

  private val typeSerializer = tableSchema.toRowType.createSerializer(new ExecutionConfig)
  // socket server should be started before running the job
  private val iterator = new SocketStreamIterator[Row](0, InetAddress.getLocalHost, typeSerializer)

  override def getTableSchema: TableSchema = tableSchema

  override def getConsumedDataType: DataType = tableSchema.toRowDataType

  override def consumeDataStream(dataStream: DataStream[Row]): DataStreamSink[_] = {
    dataStream
      .addSink(new CollectSink[Row](iterator.getBindAddress, iterator.getPort, typeSerializer))
      .name("Streaming select table sink")
      .setParallelism(1)
  }

  override def setJobClient(jobClient: JobClient): Unit = {
  }

  override def getResultIterator: util.Iterator[Row] = iterator
}

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/sinks/StreamSelectTableSink.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.planner.sinks;

import org.apache.flink.api.common.ExecutionConfig;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.typeutils.TupleTypeInfo;
import org.apache.flink.table.api.TableSchema;
import org.apache.flink.table.data.RowData;
import org.apache.flink.table.sinks.RetractStreamTableSink;
import org.apache.flink.types.Row;
import org.apache.flink.types.RowKind;

/**
 * A {@link RetractStreamTableSink} for streaming select job to collect the result to local.
 *
 * <p>{@link RowData} contains {@link RowKind} attribute which
 * can represents all kind of changes. The boolean flag is useless here,
 * only because {@link RetractStreamTableSink} requires Tuple2&lt;Boolean, T&gt; type.
 */
public class StreamSelectTableSink
		extends SelectTableSinkBase<Tuple2<Boolean, RowData>>
		implements RetractStreamTableSink<RowData> {

	public StreamSelectTableSink(TableSchema tableSchema) {
		super(tableSchema, new TupleTypeInfo<Tuple2<Boolean, RowData>>(
				Types.BOOLEAN,
				createRowDataTypeInfo(tableSchema)).createSerializer(new ExecutionConfig()));
	}

	@Override
	public TypeInformation<RowData> getRecordType() {
		return createRowDataTypeInfo(getTableSchema());
	}

	@Override
	protected Row convertToRow(Tuple2<Boolean, RowData> tuple2) {
		// convert Tuple2<Boolean, RowData> to Row
		return converter.toExternal(tuple2.f1);
	}
}
